stac_validator/stac_validator.py

```diff
@@ -0,0 +1,372 @@
+"""
+Description: Validate a STAC item or catalog against the STAC specification.
+
+Usage:
+    stac_validator <stac_file> [--version STAC_VERSION] [--threads NTHREADS] [--verbose] [--timer] [--loglevel LOGLEVEL]
+
+Arguments:
+    stac_file  Fully qualified path or url to a STAC file.
+
+Options:
+    -v, --version STAC_VERSION   Version to validate against. [default: master]
+    -h, --help                   Show this screen.
+    --threads NTHREADS           Number of threads to use. [default: 10]
+    --verbose                    Verbose output. [default: False]
+    --timer                      Reports time to validate the STAC (seconds)
+    --loglevel LOGLEVEL          Numeric level of logging to report. [default: 40]
+"""
+
+import json
+import os
+import shutil
+import tempfile
+from json.decoder import JSONDecodeError
+from timeit import default_timer
+from urllib.parse import urljoin, urlparse
+from concurrent import futures
+
+import logging
+
+import requests
+from cachetools import LRUCache
+from docopt import docopt
+from jsonschema import (
+    RefResolutionError, RefResolver, ValidationError, validate
+)
+from pathlib import Path
+
+from . stac_exceptions import VersionException
+from . stac_utilities import StacVersion
+
+logger = logging.getLogger(__name__)
+
+
+cache = LRUCache(maxsize=10)
+
+
+class StacValidate:
+    def __init__(self, stac_file, version="master", loglevel=40):
+        """
+        Validate a STAC file
+        :param stac_file: file to validate
+        :param version: github tag - defaults to master
+        """
+        logging.basicConfig(format='%(asctime)s : %(levelname)s : %(thread)d : %(message)s',
+                            datefmt='%m/%d/%Y %H:%M:%S',
+                            level=int(loglevel))
+        logging.info('STAC Validator Started.')
+        self.stac_version = version
+        self.stac_file = stac_file.strip()
+        self.dirpath = ''
+
+        self.fetch_specs(self.stac_version)
+        self.message = []
+        self.status = {
+            "catalogs": {"valid": 0, "invalid": 0},
+            "collections": {"valid": 0, "invalid": 0},
+            "items": {"valid": 0, "invalid": 0},
+            "unknown": 0,
+        }
+
+    def fetch_specs(self, version):
+        """
+        Get the versions from github. Cache them if possible.
+        :return: specs
+        """
+        geojson_key = "geojson_resolver"
+        item_key = "item-{}".format(self.stac_version)
+        catalog_key = "catalog-{}".format(self.stac_version)
+
+        if item_key in cache and catalog_key in cache:
```

Comment:
should check if `geojson_key` exists since the size of the cache is small, there's a chance it may get bumped

Reply:
I've changed cacheing to use python3 builtin LRU cache. I'll make a new PR for the cache.